src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp

```diff
@@ -0,0 +1,162 @@
+        if (!key) {
+            continue;
+        }
+        auto freeKeyScopeGuard = stdext::make_scope_exit([=]() {
+            if (freeKey) {
+                NCryptFreeObject(key);
+            }
+        });
+
+        if (spec != CERT_NCRYPT_KEY_SPEC) {
+            // We don't support old non-CNG CSPs.
+            // TODO: _log("Not CERT_NCRYPT_KEY_SPEC");
+            continue;
+        }
+
+        NCRYPT_PROV_HANDLE prov = 0;
+        DWORD type = 0, size = sizeof(prov);
+        SECURITY_STATUS err =
+            NCryptGetProperty(key, NCRYPT_PROVIDER_HANDLE_PROPERTY, PBYTE(&prov), size, &size, 0);
+        if (FAILED(err)) {
+            continue; // TODO: log.
+        }
+        if (!prov) {
+            continue;
+        }
+        auto freeProvider = stdext::make_scope_exit([=]() { NCryptFreeObject(prov); });
+
+        size = sizeof(type);
+        err = NCryptGetProperty(prov, NCRYPT_IMPL_TYPE_PROPERTY, PBYTE(&type), size, &size, 0);
+        if (FAILED(err)) {
+            continue; // TODO: log.
+        }
+
+        if ((type & (NCRYPT_IMPL_HARDWARE_FLAG | NCRYPT_IMPL_REMOVABLE_FLAG)) == 0) {
+            continue; // TODO: log.
+        }
+
+        std::wstring algo(5, 0);
+        err = NCryptGetProperty(key, NCRYPT_ALGORITHM_GROUP_PROPERTY, PBYTE(algo.data()),
+                                DWORD(algo.size() + 1) * 2, &size, 0);
+        if (FAILED(err)) {
+            continue; // TODO: log.
+        }
+        algo.resize(size / 2 - 1);
+        // TODO: use algo.starts_with(L"EC") when migrating to C++20.
+        if (algo != L"RSA" && algo.rfind(L"EC", 0) != 0) {
+            // We only support RSA and ECC algorithms.
+            continue; // TODO: log.
+        }
+
+        // Acquire the key again without the CRYPT_ACQUIRE_SILENT_FLAG as QSCD devices
+        // may fail with NTE_SILENT_CONTEXT otherwise in case they require PIN entry through the
+        // driver UI. To avoid reusing the key handle, omit CRYPT_ACQUIRE_CACHE_FLAG as well.
+        // MsCryptoApiElectronicID will take ownership of the key in the next step.
+        flags = CRYPT_ACQUIRE_COMPARE_KEY_FLAG | CRYPT_ACQUIRE_ONLY_NCRYPT_KEY_FLAG;
+        key = 0;
+        spec = 0;
+        freeKey = FALSE;
+        CryptAcquireCertificatePrivateKey(cert, flags, nullptr, &key, &spec, &freeKey);
+        if (!key) {
+            continue;
+        }
+
+        // The certificate context must also remain alive to use the key, so call
+        // CertDuplicateCertificateContext() to increment its reference count to keep it alive.
+        // MsCryptoApiElectronicID frees the certificate context itself.
+        auto eid = std::make_unique<MsCryptoApiElectronicID>(CertDuplicateCertificateContext(cert),
+                                                             std::move(certData), certType,
+                                                             algo == L"RSA", key, freeKey);
+
+        msCryptoApiElectronicIDs.emplace_back(
```

Comment:
```suggestion
        msCryptoApiElectronicIDs.push_back(
```

Reply:
`emplace_back` is preferrable according to https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-emplace.html, but there are indeed other opinions as well like https://quuxplusone.github.io/blog/2021/03/03/push-back-emplace-back/. Why do you prefer `push_back`?